torch/testing/_internal/common_utils.py

```diff
@@ -1541,7 +1541,7 @@ def random_square_matrix_of_rank(l, rank, dtype=torch.double, device='cpu'):
             s[i] = 0
         elif s[i] == 0:
             s[i] = 1
-    return u.mm(torch.diag(s)).mm(v.transpose(0, 1))
+    return u.mm(torch.diag(s).to(dtype)).mm(v.transpose(0, 1))
```

Comment:
It's unfortunate that `torch.diag` doesn't take tensor options, despite it being a tensor creation op.

Reply:
It doesn't need to be transferred, the device of `diag(s)` is `s.device()`.
`to(dtype)` is needed here because for cfloat and cdouble input, `u`and`v` are of type cfloat/cdouble while `s` is float/double to be able to multiply them `s` needs to be transformed into cfloat/cdouble. Matrix multiplication doesn't allow mixing real and complex arguments.